### ISSUE TITLE
Support for custom manifest config plugins

### ISSUE THREAD

**Issue by sir-captainmorgan21** [googleapis/release-please#1529]
**Is your feature request related to a problem? Please describe.**
We use NX to manage our Workspace. The `node-workspace` plugin isn't resulting in the correct dep tree based on our set up. We have packages that depend on others, but its not resulting in that when using the `node-workspce` plugin.
 **Describe the solution you'd like**
We'd like to be able to write our own custom plugin to handle this. If you'd like to add support for NX, that would be nice too
 **Describe alternatives you've considered**
Only other alternative was switching to another package, but we like what `release-please` offers over others like semantic-release
 **Additional context**
NONE

**Comment by sir-captainmorgan21** [comment]
Was able to get it working by adding the workspace dependencies to the package.json dependency list. Supporting NX would still be nice!

**Comment by chingor13** [comment]
I'm not familiar with NX, but if it's well known, we would accept a workspace manager for it. We have an abstract `workspace` plugin that does the heavy lifting for determining which modules need version bumps, so all that would be needed to add is a NX specific parser to build the initial dependency tree.

**Comment by daton89** [comment]
I'm also using Nx and I've experienced the same with the node-workspace plugin, I don't get the expected behaviour where for example: changes on lib A which depends on B are not detected when I make changes on B. 

@sir-captainmorgan21 can you please describe how you got it working? 

The node-workspace plugin is using the Lerna style dependency graph, while Nx resolves packages dependencies on tsconfig level, that's why release please currently doesn't understand the dependencies between packages.

**Comment by sir-captainmorgan21** [comment]
@daton89 yea, so we just explicitly list the internal packages that we are dependent on in the package's package.json... So if `packageA` depends on `packageB`, then in `packageA`'s package.json, we list packageB as a dependency like so...
```
"dependencies": [
   packageB: '^1.0.0'
]
```

**Comment by sir-captainmorgan21** [comment]
@chingor13 yea, NX is one of the most widely used project management tools, especially for any mono-repo setup with Angular. I could take a stab at implementing an official plugin for this. Is there contributing docs somewhere I can get started?

**Comment by chingor13** [comment]
I started a doc with a braindump about the design of the codebase: https://github.com/googleapis/release-please/pull/1551

**Comment by nhuray** [comment]
@sir-captainmorgan21 I just saw you started to implement an `nx-workspace` on that [branch](https://github.com/sir-captainmorgan21/release-please/tree/nx-workspace). 

Do you need any help from the community ?

**Comment by allbetter-max** [comment]
> @sir-captainmorgan21 I just saw you started to implement an `nx-workspace` on that [branch](https://github.com/sir-captainmorgan21/release-please/tree/nx-workspace).
> 
> Do you need any help from the community ?

Happy to test :+1:

**Comment by sir-captainmorgan21** [comment]
@nhuray id love some help. I haven't been able to work on it as much as I'd like with everything going on at home. Last I remember I was in a decent spot where I was just cleaning up the PR output. However the way I was testing wasn't very good. I was building the package locally and manually updating the node modules in the repo where I use release please. Would love some assistance in setting up a test bed in the release please repo itself for NX

**Comment by sir-captainmorgan21** [comment]
I think the only solution is to set up an NX project somewhere in the repo to test against. Could also mock the responses of the nx functions, but that could be fairly brittle compared to using a real NX project as NX updates happen

**Comment by daton89** [comment]
is this feature planned to be implemented?

**Comment by tamusil-cen55253** [comment]
> @daton89 yea, so we just explicitly list the internal packages that we are dependent on in the package's package.json... So if `packageA` depends on `packageB`, then in `packageA`'s package.json, we list packageB as a dependency like so...
> 
> ```
> "dependencies": [
>    packageB: '^1.0.0'
> ]
> ```

Does this work for anyone? I tried but that does not work for me. I have "node-workspace" in plugins list

**Comment by airtonix** [comment]
> > @daton89 yea, so we just explicitly list the internal packages that we are dependent on in the package's package.json... So if `packageA` depends on `packageB`, then in `packageA`'s package.json, we list packageB as a dependency like so...
> > ```
> > "dependencies": [
> >    packageB: '^1.0.0'
> > ]
> > ```
> 
> Does this work for anyone? I tried but that does not work for me. I have "node-workspace" in plugins list

I wouldn't consider this a valid approach, as previous commenters stated, that NX has a separate mechanism for detecting the dep graph.

This is for good reasons too: 

- single `package.json` is faster to install in a massive monorepo
- use the tsconfig project reference pattern that NX creates for you results in faster DX in your editor
- NX has tools to create package.jsons for "publishable" packages.
- NX isn't just for nodejs, it can be used for any language: java, c#, golang, etc

this is all known to those familiar with NX (probably not to those new it it)

So the only viable solution is one that ends up using the `@nx/devkit` api tooling to discover the dep graph.

**Comment by kperdomo1** [comment]
Hey 👋 
Is this going to be implemented? Our team would love to be able to continue using release-please but failing to properly detect the dependency graph is a deal breaker 😢

**Comment by sir-captainmorgan21** [comment]
@kperdomo1 NX has since released `nx release`. Takes more work than simply using release-please, but you can set up your CI pipeline to create a PR with the latest from `nx release --dry-run` and effectively get the same result, except using NX's dep graph

**Comment by sir-captainmorgan21** [comment]
@airtonix yup 100% agree. It was just a stop gap solution until something better was implemented. It worked for us since we were simply a node package repo

**Comment by airtonix** [comment]
> [@airtonix](https://github.com/airtonix) yup 100% agree. It was just a stop gap solution until something better was implemented. It worked for us since we were simply a node package repo

Revisting this after a long hiatus from NX. I have few questions:

- did you manually list all your packages in `release-please-config.json#packages` ? 
- can you expand on this `nx release --dry-run` comment of yours? (or links to more material that mentions nx with release-please)

**Comment by sir-captainmorgan21** [comment]
Hey everyone. If you find yourself here, I highly recommend ditching release-please and taking a look at this articlue I wrote up now that NX provides its own release management tooling called `NX Release`. https://medium.com/palmetto-engineering-and-design/fully-automated-npm-package-releases-using-nx-release-28272e8b4db3
